mvebu: clearfog: support eMMC bootable images

Josua-SR opened the pull request:
Armada 388 Clearfog * boards can boot from all eMMC partitions.
Previously openwrt only generated a single image for sdcard, literally called "sdcard".

This image is incompatible with eMMC *if loading u-boot from data partition* is intended,
because for eMMC the SoC ignores LBA-1 where u-boot is located by default.

Loading U-Boot from Data partition is optional, users may have u-boot installed to eMMC boot0/boot1 partitions themselves in the past. For these users the emmc image changes nothing.

For everyone else it is very convenient to have a single image combining u-boot and rootfs.


Therefore add supporting to to generate an additional emmc image with u-boot at LBA-4096 and appropriate free space before partition #1.

Josua-SR committed:
uboot-mvebu: add clearfog-emmc build variant

Armada 388 requires different u-boot configuration for booting from eMMC
data partition.

Add new target U-Boot/clearfog-emmc, along with required u-boot
configuration.
Required patches are included upstream since v2024.01.

Added "SD-Card" suffix to original package name to differentiate between
the variants in menuconfig.

Signed-off-by: Josua Mayer <josua@solid-run.com>

Josua-SR committed:
mvebu: add targets for generating emmc-bootable images

Bootable images for eMMC must include U-Boot binaries at different
offset, at LBA-4096 because the Boot-ROM will not search at the same
location like SD-Card.

Add Build/emmc-img* targets for generating images with u-boot at
LBA-4096, using a slightly modified copy of gen_mvebu_sdcard_img.sh.

Changes specific to gen_mvebu_emmc_img.sh are:
- move start of first partition to make space for u-boot
- change dd offset for uboot binary

Signed-off-by: Josua Mayer <josua@solid-run.com>

Josua-SR committed:
mvebu: clearfog: add emmc images

Clearfog Pro and Base can depending on installed SoM boot either from
eMMC or SD-Card. Generate additional emmc image where u-boot location
has been adapted per SoC boot-rom requirements.

Signed-off-by: Josua Mayer <josua@solid-run.com>

hauke commented on package/boot/uboot-mvebu/patches/0004-arm-mvebu-clearfog-upgrade-sata-speed-to-sata-3-6Gbp.patch:
```diff
@@ -0,0 +1,47 @@
+ /* The starting board_serdes_map reflects original Clearfog Pro usage */
+ static struct serdes_map board_serdes_map[] = {
+-	{SATA0, SERDES_SPEED_3_GBPS, SERDES_DEFAULT_MODE, 0, 0},
++	{SATA0, SERDES_SPEED_6_GBPS, SERDES_DEFAULT_MODE, 0, 0},
```
Was this maybe done intentionally? Please check the commit history of u-boot.
Maybe the chip has problems with higher speeds or some boards have problems with signal integrity. 

Does this only affect u-boot r also the Linux starting next?

Josua-SR replied:
Lrom what I saw this boot-time setting affects Linux, there appears to be no dynamic reconfiguration of the serdes speed on armada-38x in Linux.
Linux will still be able to negotiate sata-2 or sata-1 though.

U-Boot history reveals nothing, it may simply have been overlooked during initial bringup.

Josua-SR commented:
Please note the main purpose of this PR was generating special emmc image, for all clearfog boards that have emmc on SoM and can't use SD. They need special u-boot offset, or use emmc boot partitions.

However I am adding a work-in-progress for Clearfog GTR support - these boards do not have an sd-card option and use only emmc.

Josua-SR commented:
I have attempted to remove the u-boot defconfig because upstream did not accept a sperate one just for different mmc offset. Instead I added a special u-boot package with emmc configuration.

However I believe this breaks the filename used in target/linux/mvebu/image/cortexa9.mk selecting the specific u-boot with and without emmc.

Josua-SR commented:
rebased again ...

Josua-SR committed:
uboot-mvebu: add clearfog-sata build variant

Armada 388 requires different u-boot configuration for booting from
SATA.

Add new target U-Boot/clearfog-sata, using the existing
clearfog_sata_config file in upstream u-boot.

Signed-off-by: Josua Mayer <josua@solid-run.com>

Josua-SR committed:
mvebu: add targets for generating sata-bootable images

Bootable images forBootable images for sata must include a u-boot binary
with different build-time options from emmc and mmc. Boot-ROM searches
LBA-1 same as sd-card.

Add Build/sataimg* targets for generating images with sata flavour u-boot
at LBA-1 using existing gen_mvebu_sdcard_img.sh.

Signed-off-by: Josua Mayer <josua@solid-run.com>

Josua-SR committed:
mvebu: clearfog: add sata images

Clearfog Base and Pro can boot from an M.2 sata card. Generate
additional sata images including suitable u-boot binary.

Signed-off-by: Josua Mayer <josua@solid-run.com>

Josua-SR commented:
Rebased once again.

hauke commented on package/boot/uboot-mvebu/Makefile:
```diff
@@ -29,6 +29,23 @@ define U-Boot/clearfog
+endef
+
+define U-Boot/clearfog-sata
+  NAME:=SolidRun ClearFog A1
```
They all have the same name in menuconfig. Please add a postfix like:
```suggestion
  NAME:=SolidRun ClearFog A1 (SATA)
```

hauke commented on target/linux/mvebu/image/gen_mvebu_emmc_img.sh:
There are only very small changes compared to the script `target/linux/mvebu/image/gen_mvebu_sdcard_img.sh`:
```
--- target/linux/mvebu/image/gen_mvebu_sdcard_img.sh	2023-12-24 12:01:36.342489893 +0100
+++ target/linux/mvebu/image/gen_mvebu_emmc_img.sh	2024-07-21 15:44:37.475536149 +0200
@@ -42,13 +42,13 @@
 
 # create real partition table using fdisk
 printf "Creating partition table: "
-set $(ptgen -o "$OUTFILE" -h $head -s $sect -l 1024 -S 0x$SIGNATURE $ptgen_args)
+set $(ptgen -o "$OUTFILE" -h $head -s $sect -l 8192 -S 0x$SIGNATURE $ptgen_args)
 printf "Done\n"
 
 # install bootloader
 if [ -n "$BOOTLOADER" ]; then
 	printf "Writing bootloader: "
-	dd of="$OUTFILE" if="$BOOTLOADER" bs=512 seek=1 conv=notrunc 2>/dev/null
+	dd of="$OUTFILE" if="$BOOTLOADER" bs=512 seek=4096 conv=notrunc 2>/dev/null
 	printf "Done\n"
 fi
 
```

Please add the variations as configuration parameters to the existing script and use the existing one.

hauke replied:
Why is the alignment to 8MB needed?

Josua-SR replied:
> Why is the alignment to 8MB needed?

It should be 4MB, sectors of 512.

- 4MB is size of emmc boot partitions
- u-boot places environment in sectors 1920-2048

So there is plenty space left. However:
1) it is a symbolic safe gap giving space for u-boot usage.
2) It aligns with common erase-block size for MMC.

Filesystems, especially any writable rootfs should be aligned based on erase blocks on mmc media,
to improve lifetime of a device.

With overlayfs and hidden overlay partition this is also recommended, I have never checked whether fstools is behaving accordingly.

Josua-SR replied:
https://git.openwrt.org/?p=project/fstools.git;a=blob;f=libfstools/rootdisk.c;h=910899a141b82beeff4d58563723080ae4e3bfda;hb=HEAD#l18

Independent from this PR, but should (imo) be changed at some point specifically for emmc (which is hard to replace once broken).

Josua-SR replied:
> Please add the variations as configuration parameters to the existing script and use the existing one.

Okay, will do.
I think the variable should indicate the numeric offset, then the script becomes generic to block storage (sata, sd-card, emmc).

Josua-SR commented:
I addressed all requests; one in particular was difficult, using a single script for creating images.
This is because whoever wrote the command line argument parsing did so in a strange way without useful comments.

Josua-SR committed:
mvebu: gen_mvebu_sdcard_img.sh: add parameter for bootloader offset

Marvell SoCs of mvebu family load boot-code at different offsets for
different storage media.
Most notably offets searched are different for Armada 388 and eMMC vs.
sd-card data partition.

Add u-boot installation as a parameter to the sdcard image creation
script.
Keep 4MB offset for first partition, so that there is enough space for
various u-boot offsets, and for the u-boot environment (e.g. clearfog
environment at sectors 1920-2048).
Also rename the script to indicate it is no longer specific to
one medium.

Update existing ussage of the script to pass additional parameter with
value 1, same as previously hard-coded within the script.

The script usage has been further simplified by eliminating the option
of passing zero arguments. out-file, boot-loader and offset are always
required, following partitions are optional.

Signed-off-by: Josua Mayer <josua@solid-run.com>

Josua-SR commented:
rebased again, removed hackfix commit, fixed author on one commit

Neustradamus commented:
@Josua-SR: Any progress with your PR?

Josua-SR commented:
> @Josua-SR: Any progress with your PR?

As far as I am concerend it has been ready for review since July 29.
